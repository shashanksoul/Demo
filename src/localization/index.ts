import en from './translations/en.json';
import ar from './translations/ar.json';
import hi from './translations/hi.json';
import i18next from 'i18next';
import {initReactI18next} from 'react-i18next';

const languageResources = {
  en: {translation: en},
  ar: {translation: ar},
  hi: {translation: hi},
};

i18next.use(initReactI18next).init({
  compatibilityJSON: 'v3',
  lng: 'en',
  fallbackLng: 'en',
  resources: languageResources,
});

export default i18next;
