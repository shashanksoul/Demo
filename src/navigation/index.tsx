import {NavigationContainer} from '@react-navigation/native';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import React from 'react';
import {HomeScreen, LoginScreen} from '../screens';
import {RootStackParamList} from './types';

const Stack = createNativeStackNavigator<RootStackParamList>();

const RootNavigation: React.FC = () => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          options={{headerShown: false}}
          name="Login"
          component={LoginScreen}
        />
        <Stack.Screen name="Home" component={HomeScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default RootNavigation;
