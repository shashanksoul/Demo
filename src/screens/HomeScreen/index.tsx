import React, {useEffect, useState} from 'react';
import {FlatList, Text, View} from 'react-native';
import MovieItem from '../../components/MovieItem';
import style from './styles';
import {useDispatch, useSelector} from 'react-redux';
import {MovieThunks} from '../../data/redux';
import {AppState} from '../../data/redux/types';
import ProgressBar from '../../components/ProgressBar';

const HomeScreen: React.FC = () => {
  const [page, setPage] = useState(1);
  const {isLoading, movies, error, currentPage, totalPages} = useSelector(
    (state: AppState) => ({
      isLoading: state.isLoading,
      movies: state.movies,
      error: state.error,
      currentPage: state.currentMoviePage,
      totalPages: state.totalPages,
    }),
  );
  const dispatch: any = useDispatch();

  useEffect(() => {
    dispatch(MovieThunks.getMovies(page));
  }, [dispatch, page]);

  return (
    <View style={style.container}>
      <Text>{error}</Text>
      <FlatList
        numColumns={2}
        onEndReachedThreshold={0.3}
        onEndReached={() => {
          if (currentPage && totalPages && currentPage <= totalPages) {
            setPage(prevPage => prevPage + 1);
          }
        }}
        data={movies}
        keyExtractor={(item, index) => `${item.title},${index}`}
        renderItem={({item}) => <MovieItem movieDetail={item} />}
      />
      {isLoading && <ProgressBar />}
    </View>
  );
};

export default HomeScreen;
