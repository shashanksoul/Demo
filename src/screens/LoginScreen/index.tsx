import React, {useState} from 'react';
import {Image, Text, TextInput, TouchableOpacity, View} from 'react-native';
import {style} from './styles';
import {isValidEmail, isValidPassword} from '../../utils/isValid';
import {LoginScreenProps} from './types';
import {useTranslation} from 'react-i18next';
import LanguagePicker from '../../components/LanguagePicker';
import {useDispatch} from 'react-redux';
import {AuthActions} from '../../data/redux';

const LoginScreen: React.FC<LoginScreenProps> = ({navigation}) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const {t} = useTranslation();
  const dispatch = useDispatch();

  const isLoginEnabled = () => isValidEmail(email) && isValidPassword(password);
  return (
    <View style={style.container}>
      <Image
        style={style.backgroundImage}
        source={require('../../assets/images/home-bg.jpg')}
      />
      <Text style={style.title}>Demo App</Text>
      <TextInput
        testID="email-input"
        value={email}
        onChangeText={setEmail}
        keyboardType="email-address"
        style={style.input}
        placeholder={t('login.user-name')}
      />
      <TextInput
        testID="password-input"
        value={password}
        onChangeText={setPassword}
        keyboardType="visible-password"
        style={style.input}
        placeholder={t('login.password')}
      />
      <TouchableOpacity
        testID="login-button"
        disabled={!isLoginEnabled()}
        activeOpacity={0.8}
        style={
          isLoginEnabled() ? style.button : [style.button, style.buttonDisabled]
        }
        onPress={() => {
          dispatch(AuthActions.logIn());
          //navigation.replace('Home');
        }}>
        <Text
          style={
            isLoginEnabled()
              ? style.buttonText
              : [style.buttonText, style.buttonDisabledText]
          }>
          {t('login.login')}
        </Text>
      </TouchableOpacity>
      <View style={style.dropDown}>
        <LanguagePicker />
      </View>
    </View>
  );
};

export default LoginScreen;
