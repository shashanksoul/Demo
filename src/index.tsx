import React, {useEffect} from 'react';
import RootNavigation from './navigation';
import {applyMiddleware, combineReducers, createStore} from 'redux';
import {Provider} from 'react-redux';
import {thunk} from 'redux-thunk';
import {authReducer, movieReducer} from './data/redux';
import './localization';
import {updateSavedLng} from './utils/language';
import {PersistGate} from 'redux-persist/integration/react';
import {persistStore, persistReducer} from 'redux-persist';
import storage from '@react-native-async-storage/async-storage';

const persistConfig = {
  key: 'root',
  storage: storage,
  whitelist: ['authReducer'],
};

const rootReducer = combineReducers({
  movieReducer,
  authReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const middleWares = [thunk];

const store = createStore(persistedReducer, applyMiddleware(...middleWares));
const persistor = persistStore(store);
const App = () => {
  useEffect(() => {
    (async () => await updateSavedLng())();
  }, []);
  return (
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <RootNavigation />
      </PersistGate>
    </Provider>
  );
};

export default App;
