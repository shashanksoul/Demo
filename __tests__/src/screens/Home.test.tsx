import {fireEvent, render} from '@testing-library/react-native';
import {HomeScreen} from '../../../src/screens';
import React from 'react';

const setState = jest.fn();
const useStateSpy = jest.spyOn(React, 'useState');
jest.mock('react-redux', () => ({
  useSelector: jest
    .fn()
    .mockImplementationOnce(() => ({
      isLoading: false,
      movies: [],
      error: undefined,
      currentPage: 1,
      totalPages: 10,
    }))
    .mockReturnValueOnce({
      isLoading: true,
      movies: [],
      error: undefined,
      currentPage: 1,
      totalPages: 10,
    })
    .mockReturnValueOnce({
      isLoading: true,
      movies: [],
      error: 'service down',
      currentPage: 1,
      totalPages: 10,
    })
    .mockReturnValue({
      isLoading: false,
      movies: [{poster_path: 'image', title: 'name'}],
      error: undefined,
      currentPage: 1,
      totalPages: 10,
    }),
  useDispatch: jest.fn().mockReturnValue(jest.fn()),
}));

describe('Home Screen', () => {
  it('renders correctly', () => {
    const loginUI = render(<HomeScreen />);
    expect(loginUI).toMatchSnapshot();
  });
  it('progress bar visible', () => {
    const {getByTestId} = render(<HomeScreen />);
    const progressBar = getByTestId('progress-bar');
    expect(progressBar).toBeTruthy();
  });
  it('error from api', () => {
    const {getByTestId} = render(<HomeScreen />);
    const errorText = getByTestId('error-text');
    expect(errorText.children).toEqual(['service down']);
  });
  it('renders all items in flat list', () => {
    const {getAllByTestId} = render(<HomeScreen />);
    const movieItems = getAllByTestId('movie-item');
    expect(movieItems.length).toEqual(1);
  });
  it('triggers when reached end of flat list', () => {
    useStateSpy.mockImplementation(() => [1, setState]);
    const {getByTestId} = render(<HomeScreen />);
    const flatList = getByTestId('flat-list');
    fireEvent(flatList, 'onEndReached');
    expect(setState).toHaveBeenCalledWith(2);
  });
});
