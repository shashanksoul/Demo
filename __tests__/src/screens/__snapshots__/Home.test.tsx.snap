// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "black",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.3}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    testID="flat-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
